Add conversion between uncertainty types

aragilar opened the pull request:
We assume that uncertainty types can be converted via variance, similar to the coordinate system where Cartesian is used as the path to convert via.

Closes #11862.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pllim commented:
Need to add change log and fix the CI. And for future reference, please do not remove the PR checklist from template when you open a new PR. Thanks!

aragilar commented:
I've restored the PR template (sorry, didn't realise there was the checklist), and fixed flake8 and added a changelog entry. I'm guessing there should be some docs explaining how to actually use this also, but maybe they should wait until everyone's happy with the design?

aragilar commented:
Not sure why rtd is failing, the docs seem to build for me via tox.

pllim commented:
`astropy/nddata/nduncertainty.py:docstring of astropy.nddata.nduncertainty.NDUncertainty.represent_as:13: WARNING: py:obj reference target not found: other_uncert`

It fails on warning. FYI.

aragilar commented:
Ah, cool, I missed that line. Fixed now.

astrofrog commented:
I am re-milestoning to v5.1 as we've now branched v5.0.x

mwcraig commented:
Apologies to all about the delay on this. At latest I have open time in ~8 days when spring break starts.

mwcraig commented:
@saimn and I had a brief discussion in slack a couple months ago about this -- it would be nice to have both the `represent_as` method and properties for `variance` and `std` as mentioned in https://github.com/astropy/astropy/issues/10128

That can work fine with your implementation here or the implementation could happen in the `variance` and `std` properties and the `represent_as` method uses those to return a value.

It seems like the most important thing is to not convert uncertainties unnecessarily, and it looks like you are doing that at least for variance https://github.com/astropy/astropy/pull/12057/files#diff-910ef06364fad1dcf1f2d72e78f6f6ba5da5689baed44f1ab444cae751331dd7R878

@aragilar and @saimn -- can you comment here? I don't see a huge advantage of one over the other. 

Either way this PR is close to being ready to merge.

saimn commented:
Yes indeed it would be great to have the `.variance` and `.std` attributes. We could use `.variance` instead of the `_convert_to/from_variance` method used here.

aragilar commented:
I do
```python
    def _convert_to_variance(self):
        return self

    @classmethod
    def _convert_from_variance(cls, var_uncert):
        return var_uncert
```
for variance so there should be no extra conversions there (only two function calls returning the same value). I can add `.variance` and `.std` (and then have the variance and std dev classes override those properties).

I'm not sure though how `_from_variance` could be replaced by `.variance` though (`_to_variance` makes sense).

aragilar commented:
I've added `.variance` and `.std` properties, and override them where it makes sense. I've also added a short-circuit where if we're converting to the same type, simply return the existing values.

aragilar commented:
Also, let me know if I should rebase this (given main has changed a fair bit since I branched this).

aragilar commented:
The `represent_as` was suggested in https://github.com/astropy/astropy/issues/11862#issuecomment-865082529=

mwcraig commented:
@aragilar -- could you please rebase? Let's see if that fixes the failures...

aragilar commented:
I've updated the tests to better reflect the new `.std` and `.variance` changes. It looks like everything passes now (apart from the what's new check). I'll add a what's new and some usage documentation.

saimn commented:
> I'm not sure though how _from_variance could be replaced by .variance though (_to_variance makes sense).

I think the idea in #10128 was that `.variance` and `.std` would return a Numpy array directly, not an uncertainty class. This is what's done in the Gemini nddata wrapper which lead to the ticket, but I realize that was not completely clear. Since we will also have `represent_as` to get the uncertainty class, I think getting the Numpy array directly with `.variance` and `.std` makes sense, since the goal of those attributes is to simplify access to the uncertainty in the computations. And in the same idea we would also need setters (which is why I thought it could replace `_to_variance` as well). Thoughts ? (Also ping @chris-simpson, if you want to have a look :))

```
>>> nd = NDData([1, 2, 3, 4], uncertainty=VarianceUncertainty([1,1,1,1]))

>>> nd.uncertainty.represent_as(StdDevUncertainty) → returns a StdDevUncertainty object

>>> nd.uncertainty.variance → returns a Numpy array

>>> nd.uncertainty.variance *= 2 → allows to update the uncertainty by knowing the computation is done with variances, which is not possible currently with nd.uncertainty.array
```

aragilar commented:
I'm wary of having setters on the uncertainty class, as opposed to the nddata class (e.g. how would they interact with `UnknownUncertainty`), especially if it implies the object itself is replaced (e.g. standard deviation switches to variance) or there are additional conversions done (to keep the same variance class).

chris-simpson commented:
My original thought in #10128 was to have setters and properties in the `NDData` that farmed things out to the uncertainty class. So, in `NDData`:

```
@property
def variance(self):
    return None if self.uncertainty is None else self.uncertainty.variance

@variance.setter
def variance(self, value):
    if isinstance(value, numbers.Number):
        value = np.full_like(self._data, value)
    if any(value < 0):
        raise ValueError("Negative values exist in the variance array")
    # maybe check that the shape matches self._data?
    if self.uncertainty is None:
        self.uncertainty = VarUncertainty(value)
    else:
        self.uncertainty.variance = value
```

(and similarly for `std` if you want) and then, in the various `NDUncertainty` classes you have properties and setters, so in `VarUncertainty`:

```
@property
def variance(self):
    return self.array

@variance.setter
def variance(self, value):
    self.array = value
```

whereas in `StdUncertainty`:

```
@property
def variance(self):
    return self.array ** 2

@variance.setter
def variance(self, value):
    self.array = np.sqrt(value)
```

Obviously you can have `NDUncertainty` subclasses that don't return, or permit you to set, the variance because such a property is ill-defined. But the method outlined here avoids silently changing the class of the `.uncertainty` attribute and also replacs the tautology of `nd.uncertainty.variance` with `nd.variance`.

aragilar commented:
It feels like there are two somewhat related feature requests we're trying to solve:
1. Convert a uncertainty instance to a instance of a different class (https://github.com/astropy/astropy/issues/11862), independent of a `NDData` instance. For me, the purpose of this is to be able to able to take data with different uncertainty classes and make them consistent for visualisation.
2. Make it easy to modify in-place the values of the uncertainty on an `NDData` class without changing which class it is (https://github.com/astropy/astropy/issues/10128). I'm guessing the use case that #10128 gives implicitly is a analysis/reduction pipeline?

If we strip this PR back to the first commit, which just adds the conversion between uncertainty classes, and document and merge that, and work out the best behaviour for the in-place modification (which can use the conversion functions if needed) in a separate PR, would that resolve everyone's concerns?

mwcraig commented:
> If we strip this PR back to the first commit, which just adds the conversion between uncertainty classes, and document and merge that, and work out the best behaviour for the in-place modification (which can use the conversion functions if needed) in a separate PR, would that resolve everyone's concerns?

@aragilar -- I think your suggestion makes a lot of sense. Let's get this in, and I (or someone else if they want) can work up a PR to do the variance stuff on Wednesday.

aragilar commented:
I've added the what's new, and added an example to the uncertainty section in nddata. I can't find a section about `NDUncertainty` usage, is there one?

aragilar commented:
I've also got the old version (with the variance properties) at https://github.com/aragilar/astropy/tree/transform-uncertainties-old if people want to reuse it in any way.

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -933,3 +981,10 @@ def _propagate_divide(self, other_uncert, result_data, correlation):
         return 1 / value ** 2
+
+    def _convert_to_variance(self):
+        return VarianceUncertainty(self.array ** -1, unit=self.unit ** -1)
```
Is there a reason not to use `1 / self.array` here? No need to change this code, I'm just curious if there waws a performance reason or something for writing it this way.

mwcraig replied:
Same issue as https://github.com/astropy/astropy/pull/12057/files#r854225811 when `self.unit is None`

aragilar replied:
Fixed. I think the `array ** -1` was a habit I picked up when doing odd things with numpy dtypes in order to conserve the type, but I'm not sure if that's still needed.

mwcraig commented on astropy/nddata/tests/test_nduncertainty.py:
```diff
@@ -354,3 +364,24 @@ def test_assigning_uncertainty_with_bad_unit_to_parent_fails(NDClass,
+    start_uncert = UncertClass(uncert)
+    final_uncert = start_uncert.represent_as(UncertClass)
+    assert_array_equal(start_uncert.array, final_uncert.array)
+    assert start_uncert.unit == final_uncert.unit
```
Could you please add a test (either in this test function or a separate one) that checks data values and units for conversion from one uncertainty class to another?

I think the test you have here ensures that the round trip through `VarianceUncertainty` works but in principle the conversion to/from variance could have compensating errors.

Checks for converting to `VarianceUncertainty` from each of the `uncertainty_types_with_conversion_support` should be sufficient, since the test above is then basically testing conversion back.

aragilar replied:
Done

mwcraig reviewed (changes requested):
Thanks again for this, @aragilar -- it is very good and it was excellent to get the tests and documentation along with the code. I'd like to add one more test -- if you don't think it is necessary or don't have time to do it let me know. I have some time over the next ~24 to add the test.

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -748,6 +782,13 @@ def _propagate_divide(self, other_uncert, result_data, correlation):
         return value
 
+    def _convert_to_variance(self):
+        return VarianceUncertainty(self.array ** 2, unit=self.unit ** 2)
```
This fails because it is possible for `self.unit` to be `None`: https://github.com/astropy/astropy/runs/6088416154?check_suite_focus=true#step:6:1500

saimn replied:
`data` can also be None when instantiating an uncertainty object, so I guess this should also be handled here.

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -748,6 +782,13 @@ def _propagate_divide(self, other_uncert, result_data, correlation):
+
+    @classmethod
+    def _convert_from_variance(cls, var_uncert):
+        return cls(var_uncert.array ** (1/2), unit=var_uncert.unit ** (1/2))
```
Same issue as https://github.com/astropy/astropy/pull/12057/files#r854225811 when `self.unit is None`

mwcraig commented on astropy/nddata/nduncertainty.py:
```diff
@@ -933,3 +981,10 @@ def _propagate_divide(self, other_uncert, result_data, correlation):
+
+    @classmethod
+    def _convert_from_variance(cls, var_uncert):
+        return cls(var_uncert.array ** -1, unit=var_uncert.unit ** -1)
```
Same issue as https://github.com/astropy/astropy/pull/12057/files#r854225811 when `self.unit is None`

mwcraig commented:
I'm stumped as to why RTD is failing. There are warnings like those described in https://github.com/astropy/astropy/issues/12790#issuecomment-1026233920 and https://github.com/astropy/astropy/issues/12784#issue-1114043295 that are perhaps the issue since warnings are being converted to errors.

pllim commented:
You need to fix this:

`docs/whatsnew/5.1.rst:104: WARNING: py:class reference target not found: astropy.nddata.uduncertainty.NDUncertainty`

saimn commented on docs/whatsnew/5.1.rst:
```diff
@@ -95,6 +96,23 @@ effectively allowing tables within tables. This applies to |Column|,
+Uncertainty classes can be transformed into each other
+======================================================
+
+Subclasses of :class:`astropy.nddata.uduncertainty.NDUncertainty` can now be
```
To fix the RTD warning :
```suggestion
Subclasses of :class:`astropy.nddata.NDUncertainty` can now be
```

```
/home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/12057/docs/whatsnew/5.1.rst:104: WARNING: py:class reference target not found: astropy.nddata.uduncertainty.NDUncertainty
```

saimn commented on docs/whatsnew/5.1.rst:
```diff
@@ -95,6 +96,23 @@ effectively allowing tables within tables. This applies to |Column|,
+variance into the desired uncertainty. A simple example of this is::
+
+    >>> import numpy as np
+    >>> from astropy.nddata.nduncertainty import InverseVariance, StdDevUncertainty
```
We should advertise using the top level package:
```suggestion
    >>> from astropy.nddata import InverseVariance, StdDevUncertainty
```

mwcraig commented:
Thanks for figuring out the RTD issue!

saimn commented on astropy/nddata/nduncertainty.py:
```diff
@@ -395,6 +395,40 @@ def _propagate_multiply(self, other_uncert, result_data, correlation):
+        from_variance = getattr(other_uncert, "_convert_from_variance", None)
+        if from_variance is None:
+            raise TypeError(
+                f"{type(other_uncert)} does not support conversion from "
```
`other_uncert` is a class : 
```suggestion
                f"{other_uncert.__name__} does not support conversion from "
```

mwcraig commented on astropy/nddata/tests/test_nduncertainty.py:
```diff
@@ -354,3 +364,36 @@ def test_assigning_uncertainty_with_bad_unit_to_parent_fails(NDClass,
+@pytest.mark.parametrize(
+    'UncertClass', uncertainty_types_with_conversion_support
+)
+def test_conversion_to_from_variance_supported(UncertClass):
```
What I had in mind here was checking that if you start with one class and convert to variance, then the variance values are correct. Something like this:

```
@pytest.mark.parametrize(
    'UncertClass', uncertainty_types_with_conversion_support,
    'to_variance_func', (lambda x: x ** 2, lambda x: x, lambda x: 1 / x)
)
def test_conversion_to_from_variance_supported(UncertClass, to_variance_func):
    uncert = np.arange(1, 11).reshape(2, 5) * u.adu
    start_uncert = UncertClass(uncert)
    var_uncert = start_uncert.represent_as(VarianceUncertainty)
    assert_allclose(to_variance_func(start_uncert.array), var_uncert.array)
    # unit check too
```

I think what you have added does check the roundtripping but the intermediate result is still unchecked (I think).

aragilar replied:
Ah, ok, I'll add that as well.

pllim commented:
@mwcraig , I am removing the `keep-open` label. See https://docs.astropy.org/en/latest/development/workflow/maintainer_workflow.html#pull-requests for our current policy. Thanks!

Also, there is a conflict in What's New. Though if we're remilestoning, you should wait for the 5.2 page to be up before moving the entry. Thanks!

mwcraig commented:
> Though if we're remilestoning, you should wait for the 5.2 page to be up before moving the entry. Thanks!

Worst case is I'll push an additional test in the morning (Friday) so this is ready to go in 5.1

aragilar commented:
I've rebased this so there should be no conflicts, and added the extra check suggested by @mwcraig.

mwcraig commented:
I'm going to do one last rebase to resolve the conflict and then merge...

eteq commented:
I know this is just waiting a rebase, but just to drop a note here, this would be very useful for specutils (astropy/specutils#938, astropy/specutils#939), so definitely want to see it in for 5.1 🎉

mwcraig commented:
Rebased, and will be merged once CI finishes and before other PRs with `whatsnew` entries get merged 😀

pllim commented on astropy/nddata/nduncertainty.py:
```diff
@@ -748,6 +782,19 @@ def _propagate_divide(self, other_uncert, result_data, correlation):
+
+    @classmethod
+    def _convert_from_variance(cls, var_uncert):
+        new_array = None if var_uncert.array is None else \
+            var_uncert.array ** (1 / 2)
+        new_unit = None if var_uncert.unit is None else \
+            var_uncert.unit ** (1 / 2)
```
Why the `\`?
```suggestion
        new_array = None if var_uncert.array is None else var_uncert.array ** (1 / 2)
        new_unit = None if var_uncert.unit is None else var_uncert.unit ** (1 / 2)
```

pllim commented on astropy/nddata/tests/test_nduncertainty.py:
```diff
@@ -73,6 +73,16 @@ def _propagate_divide(self, data, final_data):
 ]
 
+uncertainty_types_with_conversion_support = (
+    StdDevUncertainty,
+    VarianceUncertainty,
+    InverseVariance,
+)
```
```suggestion
    StdDevUncertainty, VarianceUncertainty, InverseVariance)
```

pllim commented on astropy/nddata/tests/test_nduncertainty.py:
```diff
@@ -73,6 +73,16 @@ def _propagate_divide(self, data, final_data):
+    InverseVariance,
+)
+uncertainty_types_without_conversion_support = (
+    FakeUncertainty,
+    UnknownUncertainty,
+)
```
```suggestion
    FakeUncertainty, UnknownUncertainty)
```

pllim commented on astropy/nddata/tests/test_nduncertainty.py:
```diff
@@ -354,3 +364,39 @@ def test_assigning_uncertainty_with_bad_unit_to_parent_fails(NDClass,
         ndd.uncertainty = v
+
+
+@pytest.mark.parametrize(
+    'UncertClass', uncertainty_types_with_conversion_support
+)
```
```suggestion
@pytest.mark.parametrize('UncertClass', uncertainty_types_with_conversion_support)
```

pllim commented on astropy/nddata/tests/test_nduncertainty.py:
```diff
@@ -354,3 +364,39 @@ def test_assigning_uncertainty_with_bad_unit_to_parent_fails(NDClass,
+    assert start_uncert.unit == final_uncert.unit
+
+
+@pytest.mark.parametrize(
+    'UncertClass', uncertainty_types_without_conversion_support
+)
```
```suggestion
@pytest.mark.parametrize('UncertClass', uncertainty_types_without_conversion_support)
```

pllim commented on docs/whatsnew/5.1.rst:
```diff
@@ -142,6 +143,23 @@ In addition, the logic for auto-downloads was changed to guarantee that no matte
+    >>> import numpy as np
+    >>> from astropy.nddata import InverseVariance, StdDevUncertainty
+    >>> StdDevUncertainty(np.arange(1, 10)).represent_as(InverseVariance)
+    InverseVariance([1.        , 0.25      , 0.11111111, 0.0625    ,
+                 0.04      , 0.02777778, 0.02040816, 0.015625  ,
+                 0.01234568])
```
Is the indentation really so weird or a typo?
```suggestion
    InverseVariance([1.        , 0.25      , 0.11111111, 0.0625    ,
                     0.04      , 0.02777778, 0.02040816, 0.015625  ,
                     0.01234568])
```

pllim commented:
Azure is really congested, so ignoring. Maybe @astrofrog or @Cadair can get rid of it and use Action like they did with a few other packages already...

pllim commented:
Thanks, all!
